File: 2-matching-and-dynamic-pricing/0-typescript/src/matching.controller.ts
import { Body, Controller, Get, HttpCode, Post, Query } from "@nestjs/common"
import { MatchingService } from "./matching.service"

interface RequestBody {
    riderId: string
    lat: number
    lng: number
}

@Controller()
export class MatchingController {
    constructor(private readonly matching: MatchingService) {}

    @Get("health")
    health(): { status: string } {
        return { status: "ok" }
    }

    @Get("api/match/quote")
    async quote(
        @Query("lat") lat: string,
        @Query("lng") lng: string,
    ): Promise<{ cell: string; demand: number; supply: number; multiplier: number }> {
        return this.matching.quoteSurge(Number(lat), Number(lng))
    }

    @Post("api/match/request")
    @HttpCode(200)
    async request(@Body() body: RequestBody): Promise<Record<string, unknown>> {
        const nearest = await this.matching.findNearestDriver(Number(body.lat), Number(body.lng))
        if (nearest === null) {
            return { riderId: body.riderId, matched: false }
        }
        return {
            riderId: body.riderId,
            driverId: nearest.driverId,
            distanceMeters: nearest.distanceMeters,
            ringMeters: nearest.ringMeters,
            matched: true,
        }
    }
}

File: 2-matching-and-dynamic-pricing/1-java/src/main/java/com/starci/matching/MatchingController.java
package com.starci.matching;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
public class MatchingController {

    private final MatchingService matching;

    public MatchingController(MatchingService matching) {
        this.matching = matching;
    }

    @GetMapping("/health")
    public Map<String, Object> health() {
        return Map.of("status", "ok");
    }

    @GetMapping("/api/match/quote")
    public Map<String, Object> quote(@RequestParam double lat, @RequestParam double lng) {
        return matching.quote(lat, lng);
    }

    @PostMapping("/api/match/request")
    public Map<String, Object> request(@RequestBody MatchRequest body) {
        return matching.requestRide(body.riderId(), body.lat(), body.lng());
    }

    public record MatchRequest(String riderId, double lat, double lng) {
    }
}
